### PR TITLE
DEV/BLD: generate `requirements/*` files to simplify build

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
- Improving DX while we wait for pypa/pip#11440 / python/peps#3541, along the lines of scikit-image/scikit-image#7085
- Supersedes gh-18899
- x-ref discussion in gh-18912

#### What does this implement/fix?
As reported in https://github.com/scipy/scipy/pull/18899#issue-1806698282 and https://github.com/scipy/scipy/pull/18899#issuecomment-1652361975, the developer experience around installing dependencies via `pip` could do with some improvement, including fixing some problems in [the documentation](https://scipy.github.io/devdocs/building/index.html).

This PR adopts @mwtoews 's script from sk-image (with some slight modifications).

It also:
- removes `mypy_requirements.txt` (contents moved to the `mypy` CI workflow, otherwise superseded by `dev_requirements.txt`)
- includes `ninja` in the build requirements (but not in `pyproject.toml`)
- removes `click` from the dev requirements since it is redundant with `rich-click` there
- works around the `scikit-umfpack` circular dependency issue (https://github.com/scipy/scipy/issues/18912#issuecomment-1642432735) by commenting it out of `test_requirements.txt`. IIUC we want to keep it in `pyproject.toml`, but I haven't seen a fix suggested, so work around it for now.

The docs are adjusted accordingly.

#### Additional information
Follow-ups:
- Using `venv` on M1 Mac still seems unstable with the current instructions, will open a separate issue
- Need to fix the `scikit-umfpack` circularity

cc @bsipocz @stefanv @rgommers @AlexSzatmary @tylerjereddy

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -79,6 +80,23 @@ classifiers = [
...
+    "numpy==1.23.2; python_version=='3.11' and platform_python_implementation != 'PyPy'",
+    "numpy>=1.26.0,<1.27; python_version=='3.12'",
+    "numpy>=1.25.0; python_version>='3.9' and platform_python_implementation=='PyPy'",
+    "numpy>=2.0.0.dev0; python_version>='3.13'",
+    # used by `python dev.py build`
+    "rich-click",
+    "doit>=0.36.0",
+    "pydevtool",
+    "ninja",
+]
```
Comment (lucascolley) [inline pyproject.toml:99]: maybe these can be included in `[build-system] -> requires` (I assume they're not there for a reason?).

File: tools/generate_requirements.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import sys
+from pathlib import Path
+
+try:  # standard module since Python 3.11
+    import tomllib as toml
+except ImportError:
+    try:  # available for older Python via pip
+        import tomli as toml
+    except ImportError:
+        sys.exit("Please install `tomli` first: `pip install tomli`")
```
Comment (lucascolley) [inline tools/generate_requirements.py:13]: not sure if this should be more general / mention `mamba`/`conda`

File: mypy_requirements.txt
Comment (lucascolley) [inline mypy_requirements.txt:1]: would have to modify `mypy` CI job (or just revert this removal)

File: pyproject.toml [hunk 3]
```diff
@@ -79,6 +80,23 @@ classifiers = [
 ]
 
 [project.optional-dependencies]
+build = [
```
Comment (rgommers) [inline pyproject.toml:83]: This seems conceptually wrong, these are duplicate from the `requires` table. And `dev.py` dependencies are `dev` not `build`.

File: pyproject.toml [hunk 4]
```diff
@@ -109,10 +127,6 @@ dev = [
     "pycodestyle",
     "ruff",
     "cython-lint>=0.12.2",
-    "rich-click",
-    "click",
-    "doit>=0.36.0",
-    "pydevtool",
```
Comment (rgommers) [inline pyproject.toml:115]: please leave these here

File: dev_requirements.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# Generated via tools/generate_requirements.py.
+# Do not edit this file; modify pyproject.toml instead and run the generator.
+mypy
```
Comment (rgommers) [inline dev_requirements.txt:3]: We can unpin this if tests pass, but if we get problems with it (Mypy is terrible backwards compat wise), we should put back a pin.

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @lucascolley, this seems useful to do to address some of the confusion for virtualenv users.

>  adds `[project.optional-dependencies] -> build` to `pyproject.toml` (this may be undesired for SciPy, just copying sk-image here, I don't know if this will have unintended side-effects)

This is simply undesirable duplication, and contains several conceptual issues. Let's not do this.


> includes `ninja` which appeared to be missing from the build dependencies

It's not missing, it's a system dependency that happens to be installable from PyPI also. It should not be in `pyproject.toml`; meson-python already adds it dynamically if it's not installed through the system package manager for isolated builds. It _may_ be added to `build_requirements.txt` dynamically, that can't do much harm beyond possibly duplicate installs taking up a little bit of disk space and will address the "setting up a dev env" use case where you want non-isolated builds.

> works around the `scikit-umfpack` circular dependency issue (https://github.com/scipy/scipy/issues/18912#issuecomment-1642432735) by commenting it out of `test_requirements.txt`. IIUC we want to keep it in `pyproject.toml`, but I haven't seen a fix suggested, so work around it for now.

This sounds good. I'll need to look at breaking the circle there, it wasn't on my radar.

> do we want `default_requirements.txt`? Again, here as copied from sk-image, but probably unwanted for SciPy?.

Yes that is fine. It's the one runtime requirement

>  perhaps using a `requires/*` directory like sk-image would be better?

+1. I think it's called `requirements/` typically.

- does the casing on Python version for `numpy` need to carry through to `build_requirements.txt`?

No, `numpy==` should not be included in `build_requirements.txt`. Those entries belong in `pyproject.toml`, but not in requirements files - they should be omitted by the script, or be put in with the same `>=` specifier as in `project.dependencies`. The `==` is for building wheel for distribution on PyPI, it's unrelated to setting up a dev env. And typically there one wants the most recent rather than the oldest numpy version.

Finally one idea for further usability improvement: add an `all_requirements.txt` file like this: https://github.com/numpy/numpy/blob/main/all_requirements.txt

File: .github/workflows/linux_meson.yml [hunk 6]
```diff
@@ -118,7 +118,7 @@ jobs:
       if: matrix.python-version == '3.10'
       run: |
         # Packages that are only needed for their annotations
-        python -m pip install -r mypy_requirements.txt
+        python -m pip install mypy==1.0.0 types-psutil typing_extensions
```
Comment (lucascolley) [inline .github/workflows/linux_meson.yml:121]: I've moved the `mypy` requirements here.

File: dev_requirements.txt (hunk 5 above)
Comment (lucascolley) [reply dev_requirements.txt]: Should the pin be added dynamically to `dev_requirements.txt`? I've restored the pin to the CI workflow.

**Review by lucascolley — COMMENTED** [review]
@rgommers I think that I've addressed all of your comments - will update the docs now

**Comment by stefanv** [comment]
@lucascolley it's great to have someone review this script carefully! If you come across any conceptual issues, we'd appreciate if you can propagate changes back to skimage 🙏

**Comment by lucascolley** [comment]
> If you come across any conceptual issues, we'd appreciate if you can propagate changes back to skimage 🙏

The main conceptual issue so far is the duplication in `pyproject.toml` that Ralf pointed out, although I don't know whether that was already present in skimage before this script.

I've made some adjustments to the script to conditionally modify the requirement lists coming from `pyproject.toml` and also return those lists to be reused (e.g. we remove the `numpy` lines from `build` then append the `numpy` line from the `default` list), but that's all SciPy specific at the minute. Probably only worth making more general if this were to be upstreamed somewhere... I guess that's a call on how long PEP 735 or `--only-deps` will take.

**Comment by lucascolley** [comment]
> > includes ninja which appeared to be missing from the build dependencies
>
> It's not missing, it's a system dependency that happens to be installable from PyPI also

Should it be included in the https://scipy.github.io/devdocs/building/index.html#system-level-dependencies section of the docs then, instead of dynamically adding as I've done here?

File: doc/source/building/index.rst [hunk 7]
```diff
@@ -352,17 +352,20 @@ virtual environments:
...
+       python -m pip install -r requirements/all.txt
 
-       # Test and optional runtime dependencies
-       python -m pip install pytest pytest-xdist pytest-timeout pooch threadpoolctl asv gmpy2 mpmath hypothesis
+       # Alternatively, you can install just the dependencies for certain
+       # development tasks:
 
-       # Doc build dependencies
-       python -m pip install sphinx "pydata-sphinx-theme==0.9.0" sphinx-design matplotlib numpydoc jupytext myst-nb
+       # Build and dev dependencies (for `python dev.py {build, lint, mypy}`)
```
Comment (lucascolley) [inline doc/source/building/index.rst:361]:
```suggestion
       # Build and dev dependencies (for `python dev.py`)
```

File: doc/source/building/index.rst [hunk 8]
```diff
@@ -352,17 +352,20 @@ virtual environments:
...
+       python -m pip install -r requirements/build.txt -r requirements/dev.txt
```
Comment (lucascolley) [inline doc/source/building/index.rst:362]: Unless we want to split out the extra dev dependencies for typing and linting into a separate file like `requirements/dev-extra.txt`?

**Comment by lucascolley** [comment]
The updated docs look like this:

image

File: .github/workflows/linux_meson.yml (hunk 6 above)
Comment (rgommers) [reply .github/workflows/linux_meson.yml:121]: That looks fine to me, thanks. Resolving.

File: doc/source/building/index.rst (hunk 7 above)
Comment (rgommers) [reply doc/source/building/index.rst:361]: I'd leave this as is

File: doc/source/building/index.rst (hunk 8 above)
Comment (rgommers) [reply doc/source/building/index.rst:362]: It seems fine as is.

**Review by rgommers — CHANGES_REQUESTED** [review]
This is starting to look close. A few more things:
- `gmpy2` doesn't support Python 3.12 yet, so I think we should leave it out of the test dependencies
- `all.txt` should not contain a full copy of the requirements in the other files, but instead just `-r build.txt` etc. (see numpy's `all_requirements.txt`)

**Comment by lucascolley** [comment]
> gmpy2 doesn't support Python 3.12 yet, so I think we should leave it out of the test dependencies

Do you mean

a) remove it from `pyproject.toml`
b) just remove it from `requirements/test.txt`

**Comment by rgommers** [comment]
I meant only from `requirements/test.txt`, treating it similar to `scikit-umfpack`.

**Comment by lucascolley** [comment]
All comments addressed.

**Review by rgommers — APPROVED** [review]
LGTM now, let's give this a go - thanks Lucas!

**Pull request merged**